src/kpro_req_lib.erl

```diff
@@ -0,0 +1,431 @@
+%% against one single topic-partition. In transactional mode, set
+%% `IsolationLevel = kpro_read_uncommitted' to fetch uncommitted messages.
+-spec fetch(vsn(), topic(), partition(), offset(),
+            #{ max_wait_time => wait()
+             , min_bytes => count()
+             , max_bytes => count()
+             , isolation_level => isolation_level()
+             , session_id => kpro:int32()
+             , epoch => kpro:int32()
+             }) -> req().
+fetch(Vsn, Topic, Partition, Offset, Opts) ->
+  MaxWaitTime = maps:get(max_wait_time, Opts, timer:seconds(1)),
+  MinBytes = maps:get(min_bytes, Opts, 0),
+  MaxBytes = maps:get(max_bytes, Opts, 1 bsl 20), %% 1M
+  IsolationLevel = maps:get(isolation_level, Opts, ?kpro_read_committed),
+  SessionID = maps:get(session_id, Opts, 0),
+  Epoch = maps:get(epoch, Opts, -1),
+  Fields =
+    [{replica_id, ?KPRO_REPLICA_ID},
+     {max_wait_time, MaxWaitTime},
+     {max_bytes, MaxBytes},
+     {min_bytes, MinBytes},
+     {isolation_level, IsolationLevel},
+     {session_id, SessionID},
+     {epoch, Epoch},
+     {topics,[[{topic, Topic},
+               {partitions,
+                [[{partition, Partition},
+                  {fetch_offset, Offset},
+                  {max_bytes, MaxBytes},
+                  {log_start_offset, -1} %% irelevant to clients
+                 ]]}]]},
+     % we alwyas fetch from one single topic-partition
+     % never need to forget any
+     {forgetten_topics_data, []}
+    ],
+  make(fetch, Vsn, Fields).
+
+%% @doc Help function to construct a produce request.
+produce(Vsn, Topic, Partition, Batch) ->
+  produce(Vsn, Topic, Partition, Batch, #{}).
+
+%% @doc Help function to construct a produce request.
+%% By default, it constructs a non-transactional produce request.
+%% For transactional produce requests, below conditions should be met.
+%% 1. `Batch' arg must be be a `[map()]' to indicate magic v2,
+%%     for example: `[#{key => Key, value => Value, ts => Ts}]'.
+%%     Current system time will be taken if `ts' is missing in batch input.
+%% 2. `first_sequence' must exist in `Opts'.
+%%     It should be the sequence number for the fist message in batch.
+%%     Maintained by producr, sequence numbers should start from zero and be
+%%     monotonically increasing, with one sequence number per topic-partition.
+%% 3. `txn_ctx' (which is of spec `kpro:txn_ctx()') must exist in `Opts'
+-spec produce(vsn(), topic(), partition(),
+              batch_input(), produce_opts()) -> req().
+produce(Vsn, Topic, Partition, Batch, Opts) ->
+  RequiredAcks = required_acks(maps:get(required_acks, Opts, all_isr)),
+  Compression = maps:get(compression, Opts, ?no_compression),
+  AckTimeout = maps:get(ack_timeout, Opts, ?DEFAULT_ACK_TIMEOUT),
+  TxnCtx = maps:get(txn_ctx, Opts, false),
+  FirstSequence = maps:get(first_sequence, Opts, -1),
+  EncodedBatch =
+    case TxnCtx of
+      false ->
+        kpro_batch:encode(Batch, Compression);
+      TxnCtx ->
```

Comment:
Please prefix this TxnCtx with "_", or rewrite the case to "if TxnCtx =:= false -> ...".